### PR TITLE
fix(type-utils): fixed TypeOrValueSpecifier not accounting for scoped DT packages

### LINKED ISSUE #6698: Bug: type-utils > TypeOrValueSpecifier doesn't account for org-scoped packages
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+plugin+rule+option%22) and found none that match my proposal.
- [X] I have searched the [current rule list](https://typescript-eslint.io/rules/#supported-rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Relevant Package

ast-spec

### My proposal is suitable for this project

- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Description

From https://github.com/typescript-eslint/typescript-eslint/pull/4436#discussion_r1134628059: when checking whether a `'package'` matches in `TypeOrValueSpecifier.ts`, we check whether the declaration file name includes `specifier.package` directly:

https://github.com/typescript-eslint/typescript-eslint/blob/346fdbc1c2cfe63a6cd006c22b652d536332cefc/packages/type-utils/src/TypeOrValueSpecifier.ts#L170-L176

This is a potential issue because package names with `@` in them have a `__` replacing the `@`. And there might be other replacements I'm not remembering - haven't looked deeply into it. 

https://github.com/DefinitelyTyped/DefinitelyTyped/tree/cc222f340546ff85b7c9271f4ecbd07b9d661c94/types/11ty__eleventy-img

We should handle any name substitutions in npm packages well, and add unit tests for it in `TypeOrValueSpecifier.test.ts`.

### Fail

```typescript
n/a
```

### Pass

```typescript
n/a
```

### Additional Info

_No response_

### PR DESCRIPTION
**Pull request by marekdedic**
## PR Checklist

- [x] Addresses an existing open issue: fixes #6698
- [x] That issue was marked as [accepting prs](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aopen+is%3Aissue+label%3A%22accepting+prs%22)
- [x] Steps in [Contributing](https://typescript-eslint.io/contributing) were taken

## Overview

- Added TypeOrValueSPecifier tests for package specifiers

### CODE REVIEW

File: packages/type-utils/tests/TypeOrValueSpecifier.test.ts [hunk 1]
```diff
@@ -139,7 +139,8 @@ describe('TypeOrValueSpecifier', () => {
         .program!.getTypeChecker()
         .getTypeAtLocation(
           services.esTreeNodeToTSNodeMap.get(
-            (ast.body[0] as TSESTree.TSTypeAliasDeclaration).id,
+            (ast.body[ast.body.length - 1] as TSESTree.TSTypeAliasDeclaration)
+              .id,
```
Comment (marekdedic) [inline packages/type-utils/tests/TypeOrValueSpecifier.test.ts:143]: I actually think this fixes a minor issue that is probably present in more tests - I have taken this directly out of some other tests in this repo

File: packages/type-utils/src/TypeOrValueSpecifier.ts [hunk 2]
```diff
@@ -146,6 +146,22 @@ function typeDeclaredInFile(
...
 
+function typeDeclaredInPackage(
+  packageName: string,
+  declarationFiles: ts.SourceFile[],
+): boolean {
+  const typesPackageName =
+    '@types/' +
+    (packageName[0] === '@'
+      ? packageName.substring(1).replace('/', '__')
+      : packageName);
```
Comment (JoshuaKGoldberg) [inline packages/type-utils/src/TypeOrValueSpecifier.ts:157]: [Refactor] ✨ 

```suggestion
  const typesPackageName = '@types/' + packageName.replace(/^@(.+)\//, '$1__');
```

File: packages/type-utils/src/TypeOrValueSpecifier.ts [hunk 3]
```diff
@@ -146,6 +146,22 @@ function typeDeclaredInFile(
...
+  return declarationFiles.some(
+    declaration =>
+      declaration.fileName.includes(`node_modules/${packageName}/`) ||
+      declaration.fileName.includes(`node_modules/${typesPackageName}/`),
+  );
```
Comment (JoshuaKGoldberg) [inline packages/type-utils/src/TypeOrValueSpecifier.ts:162]: [Refactor] Just to remove the need for two iterations over the string... if you dislike this I think it's totally reasonable to ignore the suggestion though 😄 

```suggestion
  const matcher = new RegExp(
    `node_modules/(?:${packageName}|${typesPackageName})/`,
  );

  return declarationFiles.some(declaration =>
    matcher.test(declaration.fileName),
  );
```

**Review by JoshuaKGoldberg — CHANGES_REQUESTED** [review]
LGTM in general, thanks for the fast follow! Just requesting changes on 1-2 cleanups, lmk what you think?

File: packages/type-utils/src/TypeOrValueSpecifier.ts (hunk 2 above)
Comment (marekdedic) [reply packages/type-utils/src/TypeOrValueSpecifier.ts:157]: Thanks, I tweaked the regexp a bit more FYI

File: packages/type-utils/src/TypeOrValueSpecifier.ts (hunk 3 above)
Comment (marekdedic) [reply packages/type-utils/src/TypeOrValueSpecifier.ts:162]: Thanks!

**Pull request merged**